README.md
# tuya ble sdk

The tuya ble sdk encapsulates the communication protocol with Tuya Smart mobile App and implements a event scheduling abilities. The device using tuya ble sdk does not need to care about the specific communication protocol implementation details. It can be interconnected with Tuya Smart App by calling the API and call back provided by the tuya ble sdk.

This topic gives details of the components, porting instruction, SDK configuration, API description, and usage of tuya ble sdk.

## Overview of tuya ble sdk

### Framework

The following figure shows the Application framework based on tuya ble sdk:

![clip_image002.png](https://airtake-public-data.oss-cn-hangzhou.aliyuncs.com/goat/20200310/4de0b2ed979d4b5593ee00f2793ca99b.png)

- platform: the chip platform. The chip and protocol stack are maintained by the chip company.

- Port: the abstract interfaces needed by the  tuya ble sdk. You must implement them according to the chip-specific platform.

- tuya ble sdk: encapsulates the communication protocol of Tuya BLE and provides the service interface to develop Tuya BLE devices.

- Application: your application, built by using tuya ble sdk.

- Tuya sdk API: to implement BLE related management, communication, and so forth. The calls of API are based on asynchronous messages, and the result of API will be notified to the Application of device by message or call back.

- sdk config: by setting the macro in the configuration file, you can configure tuya ble sdk to different modes, for example, the general network configuration mode applicable to multi-protocol devices, Bluetooth singlepoint devices mode, ECDH key based encryption method, whether to use OS, and so forth.

- Main process function: as the engine of tuya ble sdk, to which the Application will call all the time. If the platform architecture has an OS, The tuya ble sdk will automatically create a task to run the main process based on the OS related interface provided by the port layer. If the platform does not have an OS, the device Application needs to be called circularly.

- Message or Call back: SDK sends the data of status, data, and others to device Application through call back function registered by device Application or messages.

### OS compatibility

The tuya ble sdk can run on OS based chip platform other than Linux. If an OS is used, the API requests are based on asynchronous messages. When tuya ble sdk is initialized, the SDK automatically creates a task based on '`tuya_ble_config.h` file to process the message events of the SDK, and creates a message queue to receive the responses of the Application API. The results of the API are notified to the Application of the device in the form of message, so your Application needs to create a message queue and call `tuya_ble_callback_queue_register()` after calling `tuya_ble_sdk_init()` or `tuya_ble_sdk_init_async()` to register the message queue to the SDK.

In the chip platform that has an OS, you can also configure the tuya ble sdk to process messages using the task provided by Application instead of tasks within the tuya ble sdk. By doing so, the Application must implement the outbound message interface at the port layer. 

### Event queue

The earlier event takes precedence to leave (FIFO). Event queue caches the messages sent by the Application and platform layer, the event can be API calls, data response from BLE devices, and so forth. The main process function module circularly queries the message queue and takes it out for processing.

### Directories

| **Directory**               | **Description**                                              |
| --------------------------- | ------------------------------------------------------------ |
| app                         | Stores Applications that managed by the tuya ble sdk, such as Tuya test and production module, general connection modules and so forth. |
| doc                         | Help file.                                                   |
| extern\_components          | External components, for example, the extension for security-specific algorithm. |
| port                        | The abstract interfaces which must be implemented by Applications. |
| sdk                         | The core code of the tuya ble sdk.                           |
| tuya\_ble\_config.h         | The configuration file for tuya ble sdk. However, your Application needs to create another configuration files on demand. |
| tuya\_ble\_sdk\_version.h   | The version file.                                            |
| README.md                   | A brief introduction of the tuya ble sdk.                    |
| tuya\_ble\_sdk\_version.txt | Explains what are updated for each version in Chinese.       |
| CHANGELOG.md                | Explains what are updated for each version in English.       |

## The concepts of tuya ble service

The tuya ble sdk does not provide the interfaces for initializing service. Your Application needs to implement the service characteristics defined in the following table before you initializing the SDK. Other than the services required by the tuya ble sdk, you can also define other services if needed. The initial format of broadcast data must be implemented according to the following table, otherwise the tuya ble sdk cannot work.

| **Service UUID** | **Characteristic UUID** | **Properties**                | **Security Permissions** |
| ---------------- | ----------------------- | ----------------------------- | ------------------------ |
| 1910             | 2b10                    | Notify                        | None.                    |
|                  | 2b11                    | Write,write without response. | None.                    |


### The MTU

For a better compatibility, the ATT MTU used by tuya ble sdk is 23, and the GATT MTU (ATT DATA MAX) is 20. 

### Broadcast data format

The following picture illustrates the broadcast packet format of BLE.

![](https://images.tuyacn.com/fe-static/docs/img/a48b425f-19e7-40c1-986e-11f082416b49.png)

The following table describes what are contained in the broadcast packet.

| Broadcast data segment                       | Type | Description                                                  |
| -------------------------------------------- | ---- | ------------------------------------------------------------ |
| Physical connection identifier of BLE device | 0x01 | Length: 0x02;Type: 0X01; Data: 0x16                          |
| Service UUID                                 | 0x02 | Length: 0x03; Type: 0x02; Data: 0xA201                       |

CHANGELOG.md

# Change Log



## [2.0.0] - 2020-10-04

### Added

- Added support for MTU larger than 20 bytes for adapting applications running BLE4.2 and above.
- Added a new dp data receiving and sending function, which can send and receive dp data with a length of more than 255 bytes, and supports the sending mode without response.
- Added device unbinding function, which is different from the factory reset function and will not lose cloud historical data after binding again.
- Added asynchronous response of unbinding and factory reset functions.
- Added support for BEACON KEY.

### Changed

- Changed the queue scheduling mechanism under non-RTOS systems to prevent SDK from occupying CPU for a long time.
- Modify `TUYA_BLE_GATT_SEND_DATA_QUEUE_SIZE` to be automatically configured by the SDK.

### Fixed

- Fixed a bug that the device will automatic restart when continuously calling the `tuya_ble_device_factory_reset()`function.



## [1.3.0] - 2020-08-07

### Changed

- Use the new Tuya exclusive service uuid and characteristic uuid.
- Changed advertising packets format  to adapt to the new service uuid.

### Fixed

- Changed the time point at which the application callback is executed during unbinding to prevent the application from performing a reset operation in the corresponding callback function to cause unbinding failure.



## [1.2.6] - 2020-07-13

### Added

- Added support for transmitting character dp point data with length 0.

### Changed

- Changed  the definition of the critical section interface function of the port layer.

### Fixed

- Fixed the bug that the single packet data instruction cannot be parsed (Although there is no such instruction currently).



## [1.2.5] - 2020-06-03

### Changed

- Changed the dynamic memory release logic in some event handlers, the previous method may cause the dynamic memory free space to become smaller in extreme scenarios.

### Fixed

- Fix the bug of test link failure based on Tuya protocol.



## [1.2.4] - 2020-05-22

### Changed

- Changed flash interface asynchronous.
- Changed sdk information storage function based on asynchronous flash interface.
- Modify the SDK initialization function to be an asynchronous interface.



## [1.2.3] - 2020-05-07

### Added
- Added flash interface asynchronous.
- Added sdk information storage function based on asynchronous flash interface(Currently only supports nrf52832).


## [1.2.2] - 2020-04-23
### Added
- Added on-demand connection interface.
- Added the interface to get the total size, space and number of events of the scheduling queue under no OS environment.
- Added configuration item to configure whether to use the MAC address of tuya authorization information.
- Added tuya app log interface.

### Changed
- Changed the production test code structure, and remove the application layer header file to obtain the application version number, fingerprint and other information methods.
- Changed the encryption method of ble channel used in production test to support unencrypted transmission.

### Fixed
- Fix a problem that caused a compilation error in the SEGGER Embedded Studio environment.

## [1.2.1] - 2020-03-20
### Changed

sdk/src/tuya_ble_main.c
/**
 * \file tuya_ble_main.c
 *
 * \brief
 */
/*
 *  Copyright (C) 2014-2019, Tuya Inc., All Rights Reserved
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may
 *  not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  This file is part of tuya ble sdk
 */

#include "tuya_ble_stdlib.h"
#include "tuya_ble_type.h"
#include "tuya_ble_heap.h"
#include "tuya_ble_mem.h"
#include "tuya_ble_api.h"
#include "tuya_ble_port.h"
#include "tuya_ble_main.h"
#include "tuya_ble_secure.h"
#include "tuya_ble_data_handler.h"
#include "tuya_ble_storage.h"
#include "tuya_ble_sdk_version.h"
#include "tuya_ble_event.h"
#include "tuya_ble_utils.h"
#include "tuya_ble_app_uart_common_handler.h"
#include "tuya_ble_app_production_test.h"
#include "tuya_ble_log.h"
#include "tuya_ble_internal_config.h"

tuya_ble_parameters_settings_t tuya_ble_current_para;


static volatile tuya_ble_connect_status_t tuya_ble_connect_status = UNKNOW_STATUS;

#if TUYA_BLE_USE_OS

uint32_t m_cb_queue_table[TUYA_BLE_MAX_CALLBACKS];

#else
tuya_ble_callback_t m_cb_table[TUYA_BLE_MAX_CALLBACKS];
#endif


void tuya_ble_connect_status_set(tuya_ble_connect_status_t status)
{
    tuya_ble_device_enter_critical();
    tuya_ble_connect_status = status;
    tuya_ble_device_exit_critical();

}

tuya_ble_connect_status_t tuya_ble_connect_status_get(void)
{
    return  tuya_ble_connect_status;
}





#if TUYA_BLE_USE_OS

#if TUYA_BLE_SELF_BUILT_TASK
void *tuya_ble_task_handle;   //!< APP Task handle
void *tuya_ble_queue_handle;  //!< Event queue handle

static void tuya_ble_main_task(void *p_param)
{
    tuya_ble_evt_param_t tuya_ble_evt;
    uint32_t i = 0;

    while (true)
    {
        if (tuya_ble_os_msg_queue_recv(tuya_ble_queue_handle, &tuya_ble_evt, 0xFFFFFFFF) == true)
        {
            tuya_ble_event_process(&tuya_ble_evt);

        }

    }

}
#endif

#else

void tuya_ble_main_tasks_exec(void)
{
    tuya_sched_execute();
}
#endif




void tuya_ble_event_init(void)
{
#if TUYA_BLE_USE_OS
    
#if TUYA_BLE_SELF_BUILT_TASK
    tuya_ble_os_task_create(&tuya_ble_task_handle, "tuya", tuya_ble_main_task, 0, TUYA_BLE_TASK_STACK_SIZE,TUYA_BLE_TASK_PRIORITY);
    tuya_ble_os_msg_queue_create(&tuya_ble_queue_handle, MAX_NUMBER_OF_TUYA_MESSAGE, sizeof(tuya_ble_evt_param_t));
#endif
    
#else
    tuya_ble_event_queue_init();
#endif

#if TUYA_BLE_USE_OS
    for(uint8_t i= 0; i<TUYA_BLE_MAX_CALLBACKS; i++)
    {
        m_cb_queue_table[i] = 0;
    }
     
#else
    for(uint8_t i= 0; i<TUYA_BLE_MAX_CALLBACKS; i++)
    {
        m_cb_table[i] = NULL;
    }
#endif

}


uint8_t tuya_ble_event_send(tuya_ble_evt_param_t *evt)
{
#if TUYA_BLE_USE_OS

#if TUYA_BLE_SELF_BUILT_TASK    
    if(tuya_ble_os_msg_queue_send(tuya_ble_queue_handle, evt, 0))
    {
        return 0;
    }
    else
    {
        return 1;
    }
#else
   if(tuya_ble_event_queue_send_port(evt,0))
   {
       return 0;
   }
   else
   {
       return 1;
   }   
#endif
    
#else
    if(tuya_ble_message_send(evt)==TUYA_BLE_SUCCESS)
    {
        return 0;
    }
    else
    {
        return 1;
    }
#endif
}


uint8_t tuya_ble_custom_event_send(tuya_ble_custom_evt_t evt)
{
    static tuya_ble_evt_param_t event;
    uint8_t ret = 0;

#if TUYA_BLE_USE_OS
    event.hdr.event = TUYA_BLE_EVT_CUSTOM;
    event.custom_evt = evt;
    
#if TUYA_BLE_SELT_BUILT_TASK    
    if(tuya_ble_os_msg_queue_send(tuya_ble_queue_handle, &event, 0))
    {
        return 0;
    }
    else
    {
        return 1;
    }
#else
   if(tuya_ble_event_queue_send_port(&event,0))
   {
       return 0;
   }
   else
   {
       return 1;
   }   
#endif  
    
#else
    tuya_ble_device_enter_critical();
    event.hdr.event = TUYA_BLE_EVT_CUSTOM;
    event.custom_evt = evt;

    if(tuya_ble_message_send(&event)==TUYA_BLE_SUCCESS)
    {
        ret = 0;
    }
    else
    {        
        ret = 1;
    }    
    tuya_ble_device_exit_critical();
    return ret;
#endif
}



tuya_ble_status_t tuya_ble_inter_event_response(tuya_ble_cb_evt_param_t *param)
{

    switch (param->evt)
    {
    case TUYA_BLE_CB_EVT_CONNECTE_STATUS:
        break;
    case TUYA_BLE_CB_EVT_DP_WRITE:
        if(param->dp_write_data.p_data)
        {
            tuya_ble_free(param->dp_write_data.p_data);
        }
        break;
    case TUYA_BLE_CB_EVT_DP_DATA_RECEIVED:
        if(param->dp_received_data.p_data)
        {
            tuya_ble_free(param->dp_received_data.p_data);
        }
        break; 
    case TUYA_BLE_CB_EVT_DP_QUERY:
        if(param->dp_query_data.p_data)
        {
            tuya_ble_free(param->dp_query_data.p_data);
        }
        break;
    case TUYA_BLE_CB_EVT_OTA_DATA:
        if(param->ota_data.p_data)
        {
            tuya_ble_free(param->ota_data.p_data);
        }
        break;
    case TUYA_BLE_CB_EVT_NETWORK_INFO:
        if(param->network_data.p_data)
        {
            tuya_ble_free(param->network_data.p_data);
        }
        break;
    case TUYA_BLE_CB_EVT_WIFI_SSID:
        if(param->wifi_info_data.p_data)
        {
            tuya_ble_free(param->wifi_info_data.p_data);
        }
        break;
    case TUYA_BLE_CB_EVT_TIME_STAMP:
        break;
    case TUYA_BLE_CB_EVT_TIME_NORMAL:
        break;
    case TUYA_BLE_CB_EVT_DATA_PASSTHROUGH:

        if(param->ble_passthrough_data.p_data)
        {
            tuya_ble_free(param->ble_passthrough_data.p_data);
        }
        break;
    default:
        break;
    }

    return TUYA_BLE_SUCCESS;
}



uint8_t tuya_ble_cb_event_send(tuya_ble_cb_evt_param_t *evt)
{
#if TUYA_BLE_USE_OS
    if(m_cb_queue_table[0])
    {
        if(tuya_ble_os_msg_queue_send((void *)m_cb_queue_table[0], evt, 0))
        {
            return 0;
        }
        else
        {
            return 1;
        }
    }
#else
    tuya_ble_callback_t fun;
    if(m_cb_table[0])
    {
        fun = m_cb_table[0];
        fun(evt);
        tuya_ble_inter_event_response(evt);
    }
#endif
    return 0;

}


#if (TUYA_BLE_PROTOCOL_VERSION_HIGN==4)

/** @brief  GAP - Advertisement data (max size = 31 bytes, best kept short to conserve power) */

#define TUYA_BLE_ADV_DATA_LEN_MAX  31

static const uint8_t adv_data_const[TUYA_BLE_ADV_DATA_LEN_MAX] =
{
    0x02,
    0x01,
    0x06,
    0x03,
    0x02,
    0x50, 0xFD,
    0x17,
    0x16,
    0x50, 0xFD,
    0x41, 0x00,       //Frame Control
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};


#define TUYA_BLE_SCAN_RSP_DATA_LEN_MAX  31
static const uint8_t scan_rsp_data_const[TUYA_BLE_SCAN_RSP_DATA_LEN_MAX] =
{
    0x03,
    0x09,
    0x54, 0x59,
    0x17,             // length 
    0xFF,
    0xD0,
    0x07,
    0x00, //Encry Mode(8)
    0x00,0x00, //communication way bit0-mesh bit1-wifi bit2-zigbee bit3-NB
    0x00, //FLAG
    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
};

static uint8_t adv_data[TUYA_BLE_ADV_DATA_LEN_MAX];
static uint8_t scan_rsp_data[TUYA_BLE_SCAN_RSP_DATA_LEN_MAX];

uint8_t tuya_ble_get_adv_connect_request_bit_status(void)
{
    return (adv_data[11]&0x02);
}

void tuya_ble_adv_change(void)
{
    uint8_t *aes_buf = NULL;
    uint8_t aes_key[16];
    uint8_t encry_device_id[DEVICE_ID_LEN];

    memcpy(adv_data,adv_data_const,TUYA_BLE_ADV_DATA_LEN_MAX);
    memcpy(&scan_rsp_data,scan_rsp_data_const,TUYA_BLE_SCAN_RSP_DATA_LEN_MAX);

    adv_data[7] = 7+tuya_ble_current_para.pid_len;
    
    adv_data[13] = tuya_ble_current_para.pid_type;
    adv_data[14] = tuya_ble_current_para.pid_len;
    
    adv_data[11] &=(~0x04);
            
    adv_data[11] &=(~0x02);  //clear connect request bit

    scan_rsp_data[8] = TUYA_BLE_SECURE_CONNECTION_TYPE;

    scan_rsp_data[9] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY>>8;
    scan_rsp_data[10] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY;
    
    if(tuya_ble_current_para.pid_len==20)
    {
        scan_rsp_data[11] |=0x01 ;
    }
    else
    {
        scan_rsp_data[11] &=(~0x01);
    }

    if(tuya_ble_current_para.sys_settings.bound_flag == 1)
    {
        adv_data[11] |=0x08 ;
        //
        memcpy(aes_key,tuya_ble_current_para.sys_settings.login_key,LOGIN_KEY_LEN);
        memcpy(aes_key+LOGIN_KEY_LEN,tuya_ble_current_para.auth_settings.device_id,16-LOGIN_KEY_LEN);

        aes_buf = tuya_ble_malloc(200);

        if(aes_buf==NULL)
        {
            TUYA_BLE_LOG_ERROR("Malloc failed for AES BUF in adv change.");
            return;
        }
        else
        {
            memset(aes_buf,0,200);
        }

        tuya_ble_encrypt_old_with_key(aes_key,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,aes_buf);

        memcpy(&adv_data[15],(uint8_t *)(aes_buf+1),tuya_ble_current_para.pid_len);

        tuya_ble_free(aes_buf);

        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);

    }
    else
    {
        adv_data[11] &=(~0x08);

        memcpy(&adv_data[15],tuya_ble_current_para.pid,tuya_ble_current_para.pid_len);
        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);
    }
    TUYA_BLE_LOG_INFO("adv data changed ,current bound flag = %d",tuya_ble_current_para.sys_settings.bound_flag);

    tuya_ble_gap_advertising_adv_data_update(adv_data,tuya_ble_current_para.pid_len+15);
    tuya_ble_gap_advertising_scan_rsp_data_update(scan_rsp_data,scan_rsp_data_const[0]+25);

}

void tuya_ble_adv_change_with_connecting_request(void)
{
    uint8_t *aes_buf = NULL;
    uint8_t aes_key[16];
    uint8_t encry_device_id[DEVICE_ID_LEN];

    memcpy(adv_data,adv_data_const,TUYA_BLE_ADV_DATA_LEN_MAX);
    memcpy(&scan_rsp_data,scan_rsp_data_const,TUYA_BLE_SCAN_RSP_DATA_LEN_MAX);

    adv_data[7] = 7+tuya_ble_current_para.pid_len;
    
    adv_data[13] = tuya_ble_current_para.pid_type;
    adv_data[14] = tuya_ble_current_para.pid_len;
    

    adv_data[11] &=(~0x04);
    
        
    adv_data[11] |= 0x02;  //set connect request bit


    scan_rsp_data[8] = TUYA_BLE_SECURE_CONNECTION_TYPE;

    scan_rsp_data[9] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY>>8;
    scan_rsp_data[10] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY;
    
    if(tuya_ble_current_para.pid_len==20)
    {
        scan_rsp_data[11] |=0x01 ;
    }
    else
    {
        scan_rsp_data[11] &=(~0x01);
    }

    if(tuya_ble_current_para.sys_settings.bound_flag == 1)
    {
        adv_data[11] |=0x08 ;
        //
        memcpy(aes_key,tuya_ble_current_para.sys_settings.login_key,LOGIN_KEY_LEN);
        memcpy(aes_key+LOGIN_KEY_LEN,tuya_ble_current_para.auth_settings.device_id,16-LOGIN_KEY_LEN);

        aes_buf = tuya_ble_malloc(200);

        if(aes_buf==NULL)
        {
            TUYA_BLE_LOG_ERROR("Malloc failed for AES BUF in adv change with connecting request.");
            return;
        }
        else
        {
            memset(aes_buf,0,200);
        }

        tuya_ble_encrypt_old_with_key(aes_key,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,aes_buf);

        memcpy(&adv_data[15],(uint8_t *)(aes_buf+1),tuya_ble_current_para.pid_len);

        tuya_ble_free(aes_buf);

        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);

    }
    else
    {
        adv_data[11] &=(~0x08);

        memcpy(&adv_data[15],tuya_ble_current_para.pid,tuya_ble_current_para.pid_len);
        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);
    }
    TUYA_BLE_LOG_INFO("adv data changed ,current bound flag = %d",tuya_ble_current_para.sys_settings.bound_flag);
    tuya_ble_gap_advertising_adv_data_update(adv_data,tuya_ble_current_para.pid_len+15);
    tuya_ble_gap_advertising_scan_rsp_data_update(scan_rsp_data,scan_rsp_data_const[0]+25);
    
}

#endif


#if (TUYA_BLE_PROTOCOL_VERSION_HIGN==3)

/** @brief  GAP - Advertisement data (max size = 31 bytes, best kept short to conserve power) */

#define TUYA_BLE_ADV_DATA_LEN  31

static const uint8_t adv_data_const[TUYA_BLE_ADV_DATA_LEN] =
{
    0x02,
    0x01,
    0x06,
    0x03,
    0x02,
    0x50, 0xFD,
    0x17,
    0x16,
    0x50, 0xFD,
    0x31, 0x00,       //Frame Control
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};

#define TUYA_BLE_SCAN_RSP_DATA_LEN  28
static const uint8_t scan_rsp_data_const[TUYA_BLE_SCAN_RSP_DATA_LEN] =
{
    0x03,
    0x09,
    0x54, 0x59,
    0x17,             // length 
    0xFF,
    0xD0,
    0x07,
    0x00, //Encry Mode （10）
    0x00,0x00, //communication way bit0-mesh bit1-wifi bit2-zigbee bit3-NB
    0x00, //data type
    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
};

static uint8_t adv_data[TUYA_BLE_ADV_DATA_LEN];
static uint8_t scan_rsp_data[TUYA_BLE_SCAN_RSP_DATA_LEN];

uint8_t tuya_ble_get_adv_connect_request_bit_status(void)
{
    return (adv_data[11]&0x02);
}

void tuya_ble_adv_change(void)
{
    uint8_t *aes_buf = NULL;
    uint8_t aes_key[16];
    uint8_t encry_device_id[DEVICE_ID_LEN];

    memcpy(adv_data,adv_data_const,TUYA_BLE_ADV_DATA_LEN);
    memcpy(&scan_rsp_data,scan_rsp_data_const,TUYA_BLE_SCAN_RSP_DATA_LEN);

    adv_data[7] = 7+tuya_ble_current_para.pid_len;
    
    adv_data[13] = tuya_ble_current_para.pid_type;
    adv_data[14] = tuya_ble_current_para.pid_len;
    
    adv_data[11] &=(~0x04);
           
    adv_data[11] &=(~0x02);


    scan_rsp_data[8] = TUYA_BLE_SECURE_CONNECTION_TYPE;

    scan_rsp_data[9] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY>>8;
    scan_rsp_data[10] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY;
    
    if(tuya_ble_current_para.pid_len==20)
    {
        scan_rsp_data[11] |=0x01 ;
    }
    else
    {
        scan_rsp_data[11] &=(~0x01);
    }

    if(tuya_ble_current_para.sys_settings.bound_flag == 1)
    {
        adv_data[11] |=0x08 ;
        //
        memcpy(aes_key,tuya_ble_current_para.sys_settings.login_key,LOGIN_KEY_LEN);
        memcpy(aes_key+LOGIN_KEY_LEN,tuya_ble_current_para.auth_settings.device_id,16-LOGIN_KEY_LEN);

        aes_buf = tuya_ble_malloc(200);

        if(aes_buf==NULL)
        {
            return;
        }
        else
        {
            memset(aes_buf,0,200);
        }

        tuya_ble_encrypt_old_with_key(aes_key,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,aes_buf);

        memcpy(&adv_data[15],(uint8_t *)(aes_buf+1),tuya_ble_current_para.pid_len);

        tuya_ble_free(aes_buf);

        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);

    }
    else
    {
        adv_data[11] &=(~0x08);

        memcpy(&adv_data[15],tuya_ble_current_para.pid,tuya_ble_current_para.pid_len);
        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);
    }
    TUYA_BLE_LOG_INFO("adv data changed ,current bound flag = %d",tuya_ble_current_para.sys_settings.bound_flag);
    tuya_ble_gap_advertising_adv_data_update(adv_data,tuya_ble_current_para.pid_len+15);
    tuya_ble_gap_advertising_scan_rsp_data_update(scan_rsp_data,sizeof(scan_rsp_data));

}

void tuya_ble_adv_change_with_connecting_request(void)
{
    uint8_t *aes_buf = NULL;
    uint8_t aes_key[16];
    uint8_t encry_device_id[DEVICE_ID_LEN];

    memcpy(adv_data,adv_data_const,TUYA_BLE_ADV_DATA_LEN);
    memcpy(&scan_rsp_data,scan_rsp_data_const,TUYA_BLE_SCAN_RSP_DATA_LEN);

    adv_data[7] = 7+tuya_ble_current_para.pid_len;
    
    adv_data[13] = tuya_ble_current_para.pid_type;
    adv_data[14] = tuya_ble_current_para.pid_len;
    
    adv_data[11] &=(~0x04);
            
    adv_data[11] |= 0x02;


    scan_rsp_data[8] = TUYA_BLE_SECURE_CONNECTION_TYPE;

    scan_rsp_data[9] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY>>8;
    scan_rsp_data[10] = TUYA_BLE_DEVICE_COMMUNICATION_ABILITY;
    
    if(tuya_ble_current_para.pid_len==20)
    {
        scan_rsp_data[11] |=0x01 ;
    }
    else
    {
        scan_rsp_data[11] &=(~0x01);
    }

    if(tuya_ble_current_para.sys_settings.bound_flag == 1)
    {
        adv_data[11] |=0x08 ;
        //
        memcpy(aes_key,tuya_ble_current_para.sys_settings.login_key,LOGIN_KEY_LEN);
        memcpy(aes_key+LOGIN_KEY_LEN,tuya_ble_current_para.auth_settings.device_id,16-LOGIN_KEY_LEN);

        aes_buf = tuya_ble_malloc(200);

        if(aes_buf==NULL)
        {
            return;
        }
        else
        {
            memset(aes_buf,0,200);
        }

        tuya_ble_encrypt_old_with_key(aes_key,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,aes_buf);

        memcpy(&adv_data[15],(uint8_t *)(aes_buf+1),tuya_ble_current_para.pid_len);

        tuya_ble_free(aes_buf);

        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);

    }
    else
    {
        adv_data[11] &=(~0x08);

        memcpy(&adv_data[15],tuya_ble_current_para.pid,tuya_ble_current_para.pid_len);
        tuya_ble_device_id_encrypt_v4(&adv_data[11],adv_data[7]-3,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[12],encry_device_id,DEVICE_ID_LEN);
    }
    TUYA_BLE_LOG_INFO("adv data changed ,current bound flag = %d",tuya_ble_current_para.sys_settings.bound_flag);
    tuya_ble_gap_advertising_adv_data_update(adv_data,tuya_ble_current_para.pid_len+15);
    tuya_ble_gap_advertising_scan_rsp_data_update(scan_rsp_data,sizeof(scan_rsp_data));

}

#endif

#if (TUYA_BLE_PROTOCOL_VERSION_HIGN==2)

/** @brief  GAP - scan response data (max size = 31 bytes) */
static const uint8_t scan_rsp_data_const[30] =
{
    /* Manufacturer Specific Data */
    0x1D,             /* length */
    0xFF,
    0x59,
    0x02,
    0x00,0x02,
    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
};


/** @brief  GAP - Advertisement data (max size = 31 bytes, best kept short to conserve power) */
static const uint8_t adv_data_const[20] =
{
    /* Flags */
    0x02,             /* length */
    0x01, /* type="Flags" */
    0x05,
    /* Local name */
    0x0C,             /* length */
    0x09,
    'T', 'U', 'Y', 'A', '_', 'C', 'O', 'M', 'M', 'O','N',
    0x03,
    0x02,
    0x01, 0xA2,
};



void tuya_ble_adv_change(void)
{
    static uint8_t aes_buf[200];
    uint8_t aes_key[16];
    uint8_t encry_device_id[DEVICE_ID_LEN];

    //scanrsp
    uint8_t scan_rsp_data[30];
    memcpy(&scan_rsp_data,scan_rsp_data_const,sizeof(scan_rsp_data_const));

    scan_rsp_data[4] &=(~0x70);

    if(tuya_ble_current_para.sys_settings.bound_flag == 1)
    {
        scan_rsp_data[4] |=0x80 ;
        //
        memcpy(aes_key,tuya_ble_current_para.sys_settings.login_key,LOGIN_KEY_LEN);
        memcpy(aes_key+LOGIN_KEY_LEN,tuya_ble_current_para.auth_settings.device_id,16-LOGIN_KEY_LEN);

        tuya_ble_encrypt_old_with_key(aes_key,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,aes_buf);

        memcpy(&scan_rsp_data[6],(uint8_t *)(aes_buf+1),8);

        memset(aes_key,0,sizeof(aes_key));
        memcpy(aes_key,&scan_rsp_data[6],8);

        tuya_ble_device_id_encrypt(aes_key,8,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[6+8],encry_device_id,DEVICE_ID_LEN);

    }
    else
    {
        scan_rsp_data[4] &=(~0x80);
        memcpy(&scan_rsp_data[6],tuya_ble_current_para.pid,8);
        tuya_ble_device_id_encrypt(tuya_ble_current_para.pid,8,tuya_ble_current_para.auth_settings.device_id,DEVICE_ID_LEN,encry_device_id);

        memcpy(&scan_rsp_data[6+8],encry_device_id,DEVICE_ID_LEN);
    }
    TUYA_BLE_LOG_INFO("adv data changed ,current bound flag = %d",tuya_ble_current_para.sys_settings.bound_flag);

    tuya_ble_gap_advertising_scan_rsp_data_update(scan_rsp_data,sizeof(scan_rsp_data));

}

#endif



